### PR TITLE
Support for STARTLS/STARTSSL in-band transport upgrades/renegotation

### PR DESCRIPTION
**Pull request by dirkx**
Split start_ssl_client into two phases; to allow the implementation of protocols that use some sort of in-band STARTTLS or STARTSSL signal to upgrade a plaint text connection to SSL/TLS. Examples of these protocols are XMPP, SMTP and various database TCP connections.

Addresses #9099. An example has been included as part of the documentation.

## Description of Change

Option for  Inline upgrading from a clear-text connection to an SSL/TLS connection.

   Some protocols such as SMTP, XMPP, Mysql, Postgress and others allow, or require,
   that you start the connection without encryption; and then send a command to switch
   over to encryption.

   E.g. a typical SMTP submission would entail a dialogue such as this:

   1. client connects to server in the clear
   2. server says hello
   3. client sents a EHLO
   4. server tells the client that it supports SSL/TLS
   5. client sends a 'STARTTLS' to make use of this faciltiy
   6. client/server negiotiate a SSL or TLS connection.
   7. client sends another EHLO
   8. server now tells the client what (else) is supported; such as additional authentication options.
   ... conversation continues encrypted.

   This can be enabled in WiFiClientSecure by telling it to start in plaintext:

          client.setPlainStart();

   and client is than a plain, TCP, connection (just as WiFiClient would be); until the client calls
   the method:

          client.startTLS(); // returns zero on error; non zero on success.

    After which things switch to TLS/SSL.

## Tests scenarios
Please describe on what Hardware and Software combinations you have tested this Pull Request and how.

ESP32, ESP32S

## Related links

Addresses 
#9099 and issues raised over the year in EMailSender and XMPPClient

### CODE REVIEW

File: libraries/WiFiClientSecure/src/ssl_client.cpp [hunk 1]
```diff
@@ -391,14 +398,60 @@ int send_ssl_data(sslclient_context *ssl_client, const uint8_t *data, size_t len
...
-    //log_d( "Reading HTTP response...");   //for low level debug
-    int ret = -1;
+   int ret = peek_net_receive(ssl_client,ssl_client->socket_timeout);
+   if (ret > 0)
+	ret = mbedtls_net_recv(ssl_client, data, length);
 
-    ret = mbedtls_ssl_read(&ssl_client->ssl_ctx, data, length);
+    // log_v( "%d bytes NET read of %d", ret, length);   //for low level debug
+    return ret;
+}
```
Comment (lucasssvaz) [inline libraries/WiFiClientSecure/src/ssl_client.cpp:442]: Please fix indentation

**Comment by me-no-dev** [comment]
@dirkx you need to add `.skip.esp32h2` files in the example folders

File: libraries/WiFiClientSecure/src/ssl_client.cpp (hunk 1 above)
Comment (dirkx) [reply libraries/WiFiClientSecure/src/ssl_client.cpp:442]: Is there an .ident.pro file or similar that has this projects prefered formatting ?

Comment (lucasssvaz) [reply libraries/WiFiClientSecure/src/ssl_client.cpp:442]: Not yet, but we try to keep our sources on spaces 2 or 4

File: libraries/WiFiClientSecure/examples/WiFiClientSecureProtocolUpgradeSTARTT/WiFIClientSecureProtocolUpgradeSTARTTLS.ino [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+  Serial.print("Connected to ");
+  Serial.println(ssid);
+
+  Serial.printf("\nStarting connection to server: %s:%d\n", server, submission_port);
+
+
+  // skip verification for this demo. In production one should at the very least
+  // enable TOFU; or ideally hardcode a (CA) certificate that is trusted.
+  //
```
Comment (lucasssvaz) [inline libraries/WiFiClientSecure/examples/WiFiClientSecureProtocolUpgradeSTARTT/WiFIClientSecureProtocolUpgradeSTARTTLS.ino:85]:
```suggestion
```

File: libraries/WiFiClientSecure/examples/WiFiClientSecureProtocolUpgradeSTARTT/WiFIClientSecureProtocolUpgradeSTARTTLS.ino [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+  Serial.println("Completed OK\n");
+err:
+  Serial.println("Closing connection");
+  client.stop();
+}
+
+// SMTP command repsponse start with three digits and a space;
+// or, for continuation, with three digits and a '-'.
+//
```
Comment (lucasssvaz) [inline libraries/WiFiClientSecure/examples/WiFiClientSecureProtocolUpgradeSTARTT/WiFIClientSecureProtocolUpgradeSTARTTLS.ino:137]:
```suggestion
```

File: libraries/WiFiClientSecure/examples/WiFiClientSecureProtocolUpgradeSTARTT/WiFIClientSecureProtocolUpgradeSTARTTLS.ino [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+static bool readAllSMTPLines() {
+  String s = "";
+  int i;
+
+  // blocking read; we cannot rely on a timeout
+  // of a WiFiClientSecure read; as it is non
+  // blocking.
+  //
```
Comment (lucasssvaz) [inline libraries/WiFiClientSecure/examples/WiFiClientSecureProtocolUpgradeSTARTT/WiFIClientSecureProtocolUpgradeSTARTTLS.ino:145]:
```suggestion
```

File: libraries/WiFiClientSecure/examples/WiFiClientTrustOnFirstUse/.skip.esp32h2
Comment (lucasssvaz) [inline libraries/WiFiClientSecure/examples/WiFiClientTrustOnFirstUse/.skip.esp32h2:1]: This shouldn't be here. It is from #9103

File: libraries/WiFiClientSecure/examples/WiFiClientSecureProtocolUpgradeSTARTT/WiFIClientSecureProtocolUpgradeSTARTTLS.ino
Comment (lucasssvaz) [inline libraries/WiFiClientSecure/examples/WiFiClientSecureProtocolUpgradeSTARTT/WiFIClientSecureProtocolUpgradeSTARTTLS.ino:1]: Folder and sketch name differ. I suggest to shorten the name to `WiFiClientSecureProtocolUpgrade`

**Comment by me-no-dev** [comment]
@lucasssvaz PTAL

**Comment by dirkx** [comment]
Apologies - will do - but paid customers intruding :)

File: libraries/WiFiClientSecure/examples/WiFiClientSecureProtocolUpgradeSTARTT/WiFIClientSecureProtocolUpgradeSTARTTLS.ino [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+#endif
+
+#ifndef SMTP_PORT
+#define SMTP_PORT (587) // Standard (plaintext) submission port
+#endif
+
+const char* ssid     = WIFI_NETWORK;     // your network SSID (name of wifi network)
+const char* password = WIFI_PASSWD; // your network password
+const char*  server = SMTP_HOST;  // Server URL
+const int submission_port = SMTP_PORT; // submission port.
```
Comment (P-R-O-C-H-Y) [inline libraries/WiFiClientSecure/examples/WiFiClientSecureProtocolUpgradeSTARTT/WiFIClientSecureProtocolUpgradeSTARTTLS.ino:54]: Just small formatting to look better.
```suggestion
const char* ssid = WIFI_NETWORK;       // your network SSID (name of wifi network)
const char* password = WIFI_PASSWD;    // your network password
const char* server = SMTP_HOST;        // Server URL
const int submission_port = SMTP_PORT; // submission port.
```

File: libraries/WiFiClientSecure/src/WiFiClientSecure.cpp [hunk 6]
```diff
@@ -143,16 +143,40 @@ int WiFiClientSecure::connect(const char *host, uint16_t port, const char *CA_ce
...
+        log_i("startTLS: starting TLS/SSL on this dplain connection");
+        ret = ssl_starttls_handshake(sslclient);
+        if (ret < 0) {
+                log_e("startTLS: %d", ret);
+                stop();
+                return 0;
+        };
+        _stillinPlainStart = false;
+    } else 
+        log_e("startTLS: ignoring StartTLS - as we should be secure already");
```
Comment (P-R-O-C-H-Y) [inline libraries/WiFiClientSecure/src/WiFiClientSecure.cpp:176]: Suggest changing this error message to info, as it's just ignoring to start TLS again as should be started already.
```suggestion
        log_i("startTLS: ignoring StartTLS - as we should be secure already");
```

File: libraries/WiFiClientSecure/src/WiFiClientSecure.cpp [hunk 7]
```diff
@@ -254,9 +282,10 @@ int WiFiClientSecure::read(uint8_t *buf, size_t size)
         buf++;
         peeked = 1;
     }
-    
-    int res = get_ssl_receive(sslclient, buf, size);
+        res = get_ssl_receive(sslclient, buf, size);
+
     if (res < 0) {
+        log_d("Closing connection on failed read");
```
Comment (P-R-O-C-H-Y) [inline libraries/WiFiClientSecure/src/WiFiClientSecure.cpp:288]: Change debug to error message to match same log level of message on line 242.
```suggestion
        log_e("Closing connection on failed read");
```

File: libraries/WiFiClientSecure/src/WiFiClientSecure.cpp [hunk 8]
```diff
@@ -265,12 +294,17 @@ int WiFiClientSecure::read(uint8_t *buf, size_t size)
...
+    int peeked = (_peek >= 0), res = -1;
     if (!_connected) {
         return peeked;
     }
-    int res = data_to_read(sslclient);
-    if (res < 0) {
+    res = data_to_read(sslclient);
+
+    if (res < 0 && !_stillinPlainStart) {
+        log_e("Closing connection on failed avail check");
```
Comment (P-R-O-C-H-Y) [inline libraries/WiFiClientSecure/src/WiFiClientSecure.cpp:307]:
```suggestion
        log_e("Closing connection on failed available check");
```

File: libraries/WiFiClientSecure/src/WiFiClientSecure.cpp [hunk 9]
```diff
@@ -305,7 +339,7 @@ void WiFiClientSecure::setCACert (const char *rootCA)
  {
     if (bundle != NULL)
     {
-        esp_crt_bundle_set(bundle, sizeof(bundle));
+        esp_crt_bundle_set(bundle,sizeof(bundle));
```
Comment (P-R-O-C-H-Y) [inline libraries/WiFiClientSecure/src/WiFiClientSecure.cpp:342]:
```suggestion
        esp_crt_bundle_set(bundle, sizeof(bundle));
```

File: libraries/WiFiClientSecure/src/WiFiClientSecure.h [hunk 10]
```diff
@@ -78,6 +79,17 @@ class WiFiClientSecure : public WiFiClient
     bool verify(const char* fingerprint, const char* domain_name);
     void setHandshakeTimeout(unsigned long handshake_timeout);
     void setAlpnProtocols(const char **alpn_protos);
+
+    // Certain protocols start in plain-text; and then have the client
+    // give some STARTSSL command to `upgrade' the connection to TLS
+    // or SSL. setting PlainStart to true (the default is false) enables
+    // this. It is up to the application code to then call 'startTLS()'
+    // at the right point to initialise the SSL or TLS upgrade.
+    //
```
Comment (P-R-O-C-H-Y) [inline libraries/WiFiClientSecure/src/WiFiClientSecure.h:88]: Comment fixes
```suggestion
    // Certain protocols start in plain-text; and then have the client
    // give some STARTSSL command to `upgrade' the connection to TLS
    // or SSL. Setting PlainStart to true (the default is false) enables
    // this. It is up to the application code to then call 'startTLS()'
    // at the right point to initialise the SSL or TLS upgrade.
    
```

File: libraries/WiFiClientSecure/src/ssl_client.cpp [hunk 11]
```diff
@@ -226,6 +225,9 @@ int start_ssl_client(sslclient_context *ssl_client, const IPAddress& ip, uint32_
         return -1;
     }
 
+    // Note - this check for BOTH key and cert is relied on
+    // later during cleanup.
+    //
```
Comment (P-R-O-C-H-Y) [inline libraries/WiFiClientSecure/src/ssl_client.cpp:230]:
```suggestion
    
```

File: libraries/WiFiClientSecure/src/ssl_client.cpp [hunk 12]
```diff
@@ -300,15 +309,16 @@ int start_ssl_client(sslclient_context *ssl_client, const IPAddress& ip, uint32_
...
-    if (rootCABuff != NULL) {
+    if (ssl_client->ca_cert.version) {
         mbedtls_x509_crt_free(&ssl_client->ca_cert);
     }
 
-    if (cli_cert != NULL) {
+    // we know that we always have a client cert/key pair -- and we
+    // cannot look into the prviate client_key pk struct for newer
+    //versions of mbedtls. So rely on a public field of the cert
+    // and infer that there is a key too.
```
Comment (P-R-O-C-H-Y) [inline libraries/WiFiClientSecure/src/ssl_client.cpp:319]: Typo
```suggestion
    // We know that we always have a client cert/key pair -- and we
    // cannot look into the private client_key pk struct for newer
    // versions of mbedtls. So rely on a public field of the cert
    // and infer that there is a key too.
```

**Review by P-R-O-C-H-Y — CHANGES_REQUESTED** [review]
@dirkx PTAL on my comments

**Review by lucasssvaz — APPROVED** [review]

File: libraries/WiFiClientSecure/examples/WiFiClientSecureProtocolUpgradeSTARTT/WiFIClientSecureProtocolUpgradeSTARTTLS.ino (hunk 5 above)
Comment (dirkx) [reply libraries/WiFiClientSecure/examples/WiFiClientSecureProtocolUpgradeSTARTT/WiFIClientSecureProtocolUpgradeSTARTTLS.ino]: Looks good / alignment nice.

**Comment by dirkx** [comment]
Checked all the the changes & reran our testcases (against several mail provider (proton, freedom, against sendmail-sasl/etc, cyrus-imap, XMPP, OpenLDAP and against  PosgreSQL) -- and nothing breaks. 

AFAIK nothing from my end that needs to be done / anything you folks are waiting on from me ?

**Comment by lucasssvaz** [comment]
> Checked all the the changes & reran our testcases (against several mail provider (proton, freedom, against sendmail-sasl/etc, cyrus-imap, XMPP, OpenLDAP and against PosgreSQL) -- and nothing breaks.
> 
> AFAIK nothing from my end that needs to be done / anything you folks are waiting on from me ?

This PR is all good. It is pending merge.

**Pull request merged**